sdks/java/extensions/sql/jdbc/src/test/java/org/apache/beam/sdk/extensions/sql/jdbc/BeamSqlLineIT.java

```diff
@@ -0,0 +1,302 @@
+            + "taxi_rides.payload.latitude, taxi_rides.payload.longitude from taxi_rides\n"
+            + "       WHERE taxi_rides.payload.longitude > -74.747\n"
+            + "         AND taxi_rides.payload.longitude < -73.969\n"
+            + "         AND taxi_rides.payload.latitude > 40.699\n"
+            + "         AND taxi_rides.payload.latitude < 40.720 LIMIT 2;";
+
+    fixedWindowQuery =
+        "WITH geo_cells AS (\n"
+            + "         SELECT FLOOR(taxi_rides.payload.latitude / 0.05) * 0.05 AS reduced_lat,\n"
+            + "                FLOOR(taxi_rides.payload.longitude / 0.05) * 0.05 AS reduced_lon,\n"
+            + "                taxi_rides.event_timestamp\n"
+            + "         FROM taxi_rides)\n"
+            + "       SELECT COUNT(*) as num_events,\n"
+            + "              geo_cells.reduced_lat,\n"
+            + "              geo_cells.reduced_lon, \n"
+            + "              TUMBLE_START(geo_cells.event_timestamp, INTERVAL '1' SECOND)\n"
+            + "       FROM geo_cells    \n"
+            + "       GROUP BY geo_cells.reduced_lat,\n"
+            + "                geo_cells.reduced_lon,\n"
+            + "                TUMBLE(geo_cells.event_timestamp, INTERVAL '1' SECOND)\n"
+            + "       LIMIT 2;";
+
+    slidingWindowQuery =
+        "SELECT COUNT(*) AS num_events,\n"
+            + "              SUM(taxi_rides.payload.meter_increment) as revenue,\n"
+            + "              HOP_END(\n"
+            + "                 taxi_rides.event_timestamp, \n"
+            + "                 INTERVAL '1' SECOND, \n"
+            + "                 INTERVAL '2' SECOND) as minute_end\n"
+            + "              FROM taxi_rides\n"
+            + "              GROUP BY HOP(\n"
+            + "                 taxi_rides.event_timestamp,\n"
+            + "                 INTERVAL '1' SECOND, \n"
+            + "                 INTERVAL '2' SECOND) LIMIT 2";
+
+    constructor =
+        new PubsubMessageJSONStringConstructor(
+            "ride_id",
+            "point_idx",
+            "latitude",
+            "longitude",
+            "meter_reading",
+            "meter_increment",
+            "ride_status",
+            "passenger_count");
+  }
+
+  @Test
+  public void testSelectFromPubsub() throws Exception {
+    Future<List<List<String>>> expectedResult =
+        pool.submit(
+            (Callable)
+                () -> {
+                  String[] args =
+                      buildArgs(
+                          String.format(createPubsubTableStatement, eventsTopic.topicPath()),
+                          readFromPubsub);
+
+                  ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+                  BeamSqlLine.runSqlLine(args, null, outputStream, null);
+                  return toLines(outputStream);
+                });
+
+    // Wait 10 sec to allow creating a subscription.
+    Thread.sleep(10 * 1000);
```

Comment:
This will lead to flaky tests. You should create a thread that monitors for a subscription on the topic and when it sees one publishes the results.

Reply:
I thought about the waiting for solution but eventually choose this one because e2e tests talk to real systems. Waiting for 10 sec could be flaky but monitoring the creation of subscription could also have infinite waiting issue. In order to avoid long or infinite wait for subscription,  the separate thread would require a timeout setting, which is an variant of this 10 sec sleep, but just more complicated.

I would say how flaky it is will be decided by Pubsub. If 99.99% subscription can be created within 10 second, then current idea should be ok.

Reply:
As an example of imperfect world, I have this assert 
`everyItem(IsIn.isOneOf(expectedResult.get(30, TimeUnit.SECONDS).toArray())));```, which have 30 seconds timeout on waiting Pubsub messages. It could lead to flaky tests if pubsub messages delivery has very high latency. It could happen, but if we can make the timeout bigger than 999th percentile latency of Pubsub, it then might not be very bad.

Reply:
As another example of imperfect world, I don't have timeout when waiting for messages for the public topic because message arrival rate is unknown. It essentially could lead to very long wait. So far I don't have a better solution for it.

Reply:
If you have 500 tests that pass 99.9% of the time, overall Beam tests will only pass 60% of the time. I would also guess that 99% of the time this operation will succeed in less than a second, so your test is unnecessarily slow. If you wait on a subscription:
1. Your test will be faster most of the time.
2. You can increase the timeout and reduce the flakiness of the test.
3. Your failure waiting for PubSub will be clear from how the test failed. (Timeout waiting for pubsub rather then assertion failure on the results.)

Reply:
Agree. Waiting for plus time out is better than a fixed number of time.

I will change it to this way.